csrc/id_model/id_graph.h

```diff
@@ -0,0 +1,246 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <disjoint_set.h>
+#include <ir/all_nodes.h>
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace nvfuser {
+
+using IdGroup = std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>;
+using IdGroups = VectorOfUniqueEntries<IdGroup>;
+using ExprGroup = std::shared_ptr<VectorOfUniqueEntries<Expr*>>;
+using ExprGroups = VectorOfUniqueEntries<ExprGroup>;
+
+class TORCH_CUDA_CU_API IdGraph {
+ public:
+  IdGraph() = default;
+
+  IdGraph(const IdGraph& other);
+  IdGraph(IdGraph&& other) = default;
+
+  IdGraph& operator=(const IdGraph& other);
+  IdGraph& operator=(IdGraph&& other) = default;
+
+  // Returns the disjoint IterDomain set.
+  const DisjointSets<IterDomain*>& disjointIdSets() const;
+
+  DisjointSets<IterDomain*>& disjointIdSets();
+
+  // Returns the disjoint Expr set.
+  const DisjointSets<Expr*>& disjointExprSets() const;
+
+  DisjointSets<Expr*>& disjointExprSets();
+
+  // Return if there's a group entry in the graph for this expr
+  bool hasGroup(Expr* expr) const;
+
+  // Return if there's a group entry in the graph for this id
+  bool hasGroup(IterDomain* id) const;
+
+  // Convert expr to its exprGroup, assert that it exists.
+  ExprGroup toGroup(Expr* expr) const;
+
+  // Convert iter domain to its IdGroup, assert that it exists.
+  IdGroup toGroup(IterDomain* id) const;
+
+  // Convert unique vector of expressions to unique vector of its groups
+  ExprGroups toGroups(const VectorOfUniqueEntries<Expr*>& exprs) const;
+
+  // Convert unique vector of IterDomain to unique vector of its groups
+  IdGroups toGroups(const VectorOfUniqueEntries<IterDomain*>& ids) const;
+
+  // Return output/input iter domain groups of provided expr
+  std::vector<IdGroup> outputGroups(ExprGroup expr) const;
+  std::vector<IdGroup> inputGroups(ExprGroup expr) const;
+
+  // Traverses uses of the IdGroups in 'of' and returns all ExprGroups
+  // that have a use in their definition of provided of IdGroups.
+  ExprGroups allUsesOf(const IdGroups& of) const;
+
+  // Traverses definitions of the IdGroups in 'of' and returns all ExprGroups
+  // used in this history of defining the 'of' IdGroups.
+  ExprGroups allDefinitionsOf(const IdGroups& of) const;
+
+  // Return sorted expressions to go from the provided IterDomains in from to
```

Comment:
I think the term `sorted` only makes sense for DAG. If it is not a DAG, how to define `sorted`?